articles/react-error-handling.md

```diff
@@ -0,0 +1,436 @@
+
+Done that, we will create a new component called ErrorFallback and this one will receive some props passed by ErrorBoundary.
+
+```tsx
+function ErrorFallback ({ error }: ErrorFallbackProps) {
+  const { resetBoundary } = useErrorBoundary()
+
+  return (
+    <div>
+      <p>There was an error: <strong>{error.message}</strong></p>
+      <button onClick={resetBoundary}>Try Again</button>
+    </div>
+  )
+}
+```
+
+We are showing the exact thrown error message but if you would rather show another one you could do it. We are also executing a customHook to get a reset function and this is connected with a function that we can pass directly in the component ErrorBoundary like this.
+
+```tsx
+<ErrorBoundary FallbackComponent={ErrorFallback} onReset={getProductData}>
+  {renderContent()}
+</ErrorBoundary>
+```
+
+The function getProductData is the result of refactoring the useEffect:
+
+```tsx
+const getProductData = useCallback(() => {
+  if (!productId) {
+    return
+  }
+
+  setStatus('pending')
+  makeRequest(productId)
+    .then((data) => {
+      setStatus('resolved')
+      setProductData(data)
+    })
+    .catch((error: Error) => {
+      setStatus('rejected')
+      setError(error.message)
+    })
+}, [productId])
+
+React.useEffect(() => {
+  getProductData()
+}, [getProductData])
+```
+
+We isolated the logic of making the request and updating the status in a function so we can re use it inside the useEffect and outside as we do passing it as prop. By the way, the function is wrapped in an useCallback to be only updated in case productId changed.
+
+It's time to test our application:
+
+![product-search-error-boundary-try-again.gif](/images/articles/react-error-handling/product-search-error-boundary-try-again.gif)
+
+It works fine, every time we got and error the Fallback component is shown and we could execute the reset function that makes the request once again then the components passed as chidren are re mounted.
+
+Let's analyze the following outcome: 
+
+![product-search-error-boundary-reset.gif](/images/articles/react-error-handling/product-search-error-boundary-reset.gif)
+
+Fallback component is shown but hitting the search button does not do anything looking at our button's handler we will notice we are taking the inputRef's value and setting it in the state productId. Well ErrorBoundary has an approach passing the prop `resetKeys`, its value will be a key (it could be our productId) and every time this key changes ErrorBoundary will reset its internal state. There is something we should keep in mind but let's figure it out by implement it and then analyze it:
```

Comment:
```suggestion
Fallback component is shown but hitting the search button does not do anything, looking at our button's handler we will notice we are taking the inputRef's value and setting it in the state productId. Well ErrorBoundary has an approach passing the prop `resetKeys`, its value will be a key (it could be our productId) and every time this key changes, ErrorBoundary will reset its internal state. There is something we should keep in mind but let's figure it out by implementing and then analyzing it:
```